internal/datasources/service/service.go

```diff
@@ -366,7 +366,10 @@ func (d *dataSourceService) Delete(
 	tx := stx.Q()
 
 	// List rule types referencing the data source
-	ret, err := tx.ListRuleTypesReferencesByDataSource(ctx, id)
+	ret, err := tx.ListRuleTypesReferencesByDataSource(ctx, db.ListRuleTypesReferencesByDataSourceParams{
+		DataSourcesID: id,
+		ProjectID:     project,
```

Comment:
The rule type's project might not be the same as the data source's project. In this case, you want to take the hierarchy into account.